assign_rights/assemble.py

```diff
@@ -17,9 +23,27 @@ def retrieve_rights(self, rights_ids):
             rights_shells.append(RightsShell.objects.get(pk=ident))
         return rights_shells
 
-    def calculate_dates(self, end_date):
-        """docstring for calculate_dates"""
-    pass
+    def calculate_dates(self, object, request_start_date, request_end_date):
+        """Calculate rights end dates for a given object based on whether the end
+        date is open, and then based on the period and end_dates in the object.
+
+        If no end date in the object, use the grouping end date.
+        """
+        start_period = object.start_date_period
+        end_period = object.end_date_period
+        if not object.start_date:
+            object_start = request_start_date + relativedelta(years=start_period)
+        if object.start_date:
+            object_start = object.start_date + relativedelta(years=start_period)
```

Comment:
So again you've got an opportunity to factor out common logic, here and lines 40-43.

Reply:
@helrond I'm not sure if my changes are what you were going for, but I mixed it up so that we just automatically use the request dates and then change them based on some leaner logic.

Reply:
@p-galligan What I was saying was that the logic in lines 34-37 and 40-43 is exactly the same, the only difference is which attribute is being targeted. I think you can factor that out into a separate function, and then just call it twice with the relevant arguments.

Also, it looks like you're defining `end_period` twice, once in line 33 and then line 39.

Reply:
Ok, I think I've worked out the easiest way to make this work. Let me know your thoughts.